README.md
# apify-etl-cli
Provides a set of ETL CLI commands to extract, transform, and load data into funtofu database.

## Help
Type command or subcommand with -h or --help switch will display the command's usage. 

Example: `fcli -h`

## Commands
### fcli
```bash
Usage: fcli [options] [command]

Options:
  -V, --version  output the version number
  -h, --help     output usage information

Commands:
  scrape         Scrapes the web to generate dataset items in apify cloud storage
  retrieve       Retrieves last successful run dataset items from apify
  normalize      Normalizes dataset items
  load           Load normalized dataset items into database
  archive        Archives data and downloaded files
  help [cmd]     display help for [cmd]
```
### fcli scrape
```bash
Usage: fcli-scrape [options]

Retrieves last run dataset items

Options:
  -s, --silent                                    Set log level to warn (default: false)
  -d, --debug                                     Set log level to debug (default: false)
  --input-json-file <inputJsonFile>               Set the filename of the apify input config json file (default: "input.json")
  --dry-run                                       True by default, set dry run to only run against a subset of the target site (default: true)
  --requests-per-day <requestsPerDay>             Set number of requests per day to scrape (default: 500)
  --request-depths-per-day <requestDepthsPerDay>  Set number of levels of requests deep per day (default: 5)
  --daily                                         False by default, set daily flag to run against today date (default: false)
  -S, --start-data <startData>                    Set start date to scrape (default: "08/20/2019")
  -E, --end-date <endDate>                        Set the end date to scrape (default: "08/20/2019")
  --run-task-endpoint <runTaskEndpoint>           Set the apify run task endpoint (default: "https://api.apify.com/v2/actor-tasks/RUN_TASK_ENDPOINT")
  -h, --help                                      output usage information
  ```

### fcli retrieve
```bash
Usage: fcli-retrieve [options]

Retrieves last run dataset items

Options:
  -s, --silent                                 Set log level to warn (default: false)
  -d, --debug                                  Set log level to debug (default: false)
  --raw-data-dir <rawDataDir>                  Set the raw data directory to archive (default: "raw")
  --get-dataset-endpoint <getDatasetEndpoint>  Set the apify endpoint to retrieve last successful run dataset items (default: "https://api.apify.com/v2/actor-tasks/GET_DATASET_ENDPOINT")
  --data-file <dataFile>                       Set the filename to store the data file to
  -h, --help                                   output usage information
```

### fcli normalize
```bash
Usage: fcli-normalize [options]

Normalizes data extracted with scraper. Normalizer will parse data json files and populate meaningful data into the transformed data json.

Options:
  -s, --silent                                                Set log level to warn (default: false)
  -d, --debug                                                 Set log level to debug (default: false)
  --download-dir <downloadDir>                                Set the download directory to archive (default: "attachs")
  --raw-data-dir <rawDataDir>                                 Set the raw data directory to archive (default: "raw")
  --normalized-data-dir <normalizedDataDir>                   Set the normalized data directory to archive (default: "normalized")
  -c, --max-concurrency <maxConcurrency>                      Set max concurrent async requests to process at a time (default: 10)
  --life-list-thumbnail-width <lifeListThumbnailWidth>        Set the width of life list thumbnail (default: 120)
  --list-list-thubmnail-height <lifeListThumbnailHeight>      Set the height of life list thumbnail (default: 77)
  --life-detail-thumbnail-width <lifeDetailThumbnailWidth>    Set the width of life detail thumbnail (default: 400)
  --life-detail-thumbnail-height <lifeDetailThumbnailHeight>  Set the height of life detail thumbnail (default: 300)
  -D, --development                                           Set development mode to true (default: true)
  --data-file <dataFile>                                      Set the data file to process in development (default: "example.json")
  -h, --help                                                  output usage information
```

### fcli load
```bash
Usage: fcli-load [options]

Retrieves last run dataset items

Options:
  -s, --silent                               Set log level to warn (default: false)
  -d, --debug                                Set log level to debug (default: false)
  -H, --host <host>                          Set database host (default: "localhost")
  -u, --user <user>                          Set database user (default: "root")
  -p, --password <password>                  Set database password (default: "")
  -D, --database <database>                  Set database name (default: "mydatabase")
  --normalized-data-dir <normalizedDataDir>  Set the normalized data directory to archive (default: "normalized")
  -h, --help                                 output usage information
```

### fcli archive
```bash
Usage: fcli-archive [options]

Retrieves last run dataset items

Options:
  -s, --silent                               Set log level to warn (default: false)
  -d, --debug                                Set log level to debug (default: false)
  --download-dir <downloadDir>               Set the download directory to archive (default: "download")
  --raw-data-dir <rawDataDir>                Set the raw data directory to archive (default: "raw")
  --normalized-data-dir <normalizedDataDir>  Set the normalized data directory to archive (default: "normalized")
  --archived-dir <archivedDir>               Set the destination archived directory (default: "archived")
  -h, --help                                 output usage information
```

CHANGELOG.md
#### 1.1.7 (2019-09-03)

##### Other Changes

*  update scrape command default options ([5fd4a768](https://github.com/wchen02/apify-etl-cli/commit/5fd4a7688756b3acddd504f321a2d048a226f455))

#### 1.1.6 (2019-09-03)

##### Bug Fixes

*  Fix scrape command not working ([f3a34819](https://github.com/wchen02/apify-etl-cli/commit/f3a348190febbea093a815c116e4480ffabbd331))

#### 1.1.5 (2019-09-02)

##### Other Changes

*  Add input.json required for scrape command ([ae01daf3](https://github.com/wchen02/apify-etl-cli/commit/ae01daf3dbbcc8ae40de58a460db79e19ec025f5))

#### 1.1.4 (2019-09-02)

##### Chores

*  Update cli version ([9e81372c](https://github.com/wchen02/apify-etl-cli/commit/9e81372c17b36b914621f98bfce6e7a9c820975e))

##### Other Changes

*  add cpanel deployment script ([dc6ba44b](https://github.com/wchen02/apify-etl-cli/commit/dc6ba44b2bfebee12c3aa0fd22533a37e7dde4d5))

##### Refactors

*  Update default scrape options ([2c0e06a8](https://github.com/wchen02/apify-etl-cli/commit/2c0e06a8bc23cdbec5d3d2caf29473055a169038))

#### 1.1.3 (2019-08-31)

##### Other Changes

*  Update package ([d24375c4](https://github.com/wchen02/apify-etl-cli/commit/d24375c429e462dcca5e4b615b0e47107745164c))

#### 1.1.2 (2019-08-31)

##### Chores

*  Update packages ([152401e8](https://github.com/wchen02/apify-etl-cli/commit/152401e8682175c9d527df61e2bb2f37e47b4cdb))

#### 1.1.1 (2019-08-31)

##### Chores

*  Update package ([b1b80a54](https://github.com/wchen02/apify-etl-cli/commit/b1b80a54d1100e898a097766fde5ccfeedf249df))

### 1.1.0 (2019-08-31)

##### New Features

*  Added new options to fcli archive command ([4e8b79a8](https://github.com/wchen02/apify-etl-cli/commit/4e8b79a84f4e9de47ad5dbbe17308f3437762731))

#### 1.0.2 (2019-08-25)

##### Bug Fixes

*  Copy pasted bug ([bc8d95a8](https://github.com/wchen02/apify-etl-cli/commit/bc8d95a84ea984e320ab0cf919fd377dc9da36bc))

#### 1.0.1 (2019-08-25)

##### Chores

*  Update dependencies ([4d300f9c](https://github.com/wchen02/apify-etl-cli/commit/4d300f9cab739a7596ac3ebd8b4a005c13d39dd7))

## 1.0.0 (2019-08-20)

##### Documentation Changes

* **readme:**  Add help and command usages ([31b3c34d](https://github.com/wchen02/apify-etl-cli/commit/31b3c34ded479baa68ea11198c078160b6409fc6))

##### New Features

*  Create binary and make fcli command executable ([3caac24d](https://github.com/wchen02/apify-etl-cli/commit/3caac24d5c338a1ef29a78f92b9d674c6e5e96c9))
*  Add scrape CLI command ([31b57f3f](https://github.com/wchen02/apify-etl-cli/commit/31b57f3f921e3b024884b15f4f777b001003c3af))
*  Add retrieve CLI command ([f8a31732](https://github.com/wchen02/apify-etl-cli/commit/f8a317328a1b8dcaadd67fe971fec121b57a924c))
*  Add load CLI command ([1dfc3b5e](https://github.com/wchen02/apify-etl-cli/commit/1dfc3b5eb06531c863515856ea76bd048b300681))
*  Add normalize CLI command ([0d71be8f](https://github.com/wchen02/apify-etl-cli/commit/0d71be8f81d75f4acd800368f606c6a0c9c7bccb))

##### Bug Fixes

* **scrape:**  Forgot to load input json file ([9932e9d3](https://github.com/wchen02/apify-etl-cli/commit/9932e9d38a0fdd317a03b57f016b3e680b255868))
* **load:**  Change shorthand flag to avoid collision with help flag ([71aabb38](https://github.com/wchen02/apify-etl-cli/commit/71aabb38a1708b3efde677d6c4ae30d022ea6b6e))
*  Fix wrong options for normalize command ([7cff4dec](https://github.com/wchen02/apify-etl-cli/commit/7cff4deca544a89c8c67477777999908b606ece8))

##### Other Changes

*  update package ([54563b0e](https://github.com/wchen02/apify-etl-cli/commit/54563b0e0830050480e3ae422af392ba150230e3))

### 0.1.0 (2019-08-19)

##### New Features

*  Add archive command ([844bf101](https://github.com/wchen02/apify-etl-cli/commit/844bf101086c2f6f07d2b36ef38b09b3f4ecf0e3))

##### Other Changes

*  Add release scripts ([67b24d23](https://github.com/wchen02/apify-etl-cli/commit/67b24d23d8ed779e0d765bb64b70cb98767c1ddd))



fcli-load.js
const log = require('loglevel');
const program = require('commander');
const apify = require('apify-etl-lib')();

const defaultOptions = {
    silent: false,
    debug: false,
    host: 'localhost',
    user: 'root',
    password: '',
    database: 'mydatabase',
    normalizedDataDir: 'normalized',
};

program
    .description('Retrieves last run dataset items')
    .option('-s, --silent', 'Set log level to warn', defaultOptions.silent)
    .option('-d, --debug', 'Set log level to debug', defaultOptions.debug)
    .option('-H, --host <host>', 'Set database host', defaultOptions.host)
    .option('-u, --user <user>', 'Set database user', defaultOptions.user)
    .option('-p, --password <password>', 'Set database password', defaultOptions.password)
    .option('-D, --database <database>', 'Set database name', defaultOptions.database)
    .option('--normalized-data-dir <normalizedDataDir>', 'Set the normalized data directory to archive', defaultOptions.normalizedDataDir)
    .parse(process.argv)

async function main() {
    log.setLevel(getLogLevel());

    const options = {
        logLevel: getLogLevel(),
        db: {
            host: program.host,
            user: program.user,
            password: program.password,
            database: program.database,
        },
        dataDir: program.normalizedDataDir,
    };

    await apify.load(options);
}

function getLogLevel() {
    if (program.silent) {
        return log.levels.WARN;
    } else if (program.debug) {
        return log.levels.DEBUG;
    } else {
        return log.levels.INFO;
    }
}

main();


fcli-scrape.js
const log = require('loglevel');
const program = require('commander');
const apify = require('apify-etl-lib')();
const format = require('date-fns/format');
const jsonfile = require('jsonfile');

const today = new Date();

const defaultOptions = {
    silent: false,
    debug: false,
    inputJsonFile: 'input.json',
    dryRun: false,
    requestsPerDay: 5000,
    requestDepthsPerDay: 15,
    daily: false,
    startData: format(today, 'MM/DD/YYYY'),
    endDate: format(today, 'MM/DD/YYYY'),
    runTaskEndpoint: 'https://api.apify.com/v2/actor-tasks/RUN_TASK_ENDPOINT'
};

program
    .description('Retrieves last run dataset items')
    .option('-s, --silent', 'Set log level to warn', defaultOptions.silent)
    .option('-d, --debug', 'Set log level to debug', defaultOptions.debug)
    .option('--input-json-file <inputJsonFile>', 'Set the filename of the apify input config json file', defaultOptions.inputJsonFile)
    .option('--dry-run', 'Set dry run to only run against a subset of the target site', defaultOptions.dryRun)
    .option('--requests-per-day <requestsPerDay>', 'Set number of requests per day to scrape', defaultOptions.requestsPerDay)
    .option('--request-depths-per-day <requestDepthsPerDay>', 'Set number of levels of requests deep per day', defaultOptions.requestDepthsPerDay)
    .option('--daily', 'Set daily flag to run against today date', defaultOptions.daily)
    .option('-S, --start-data <startData>', 'Set start date to scrape', defaultOptions.startData)
    .option('-E, --end-date <endDate>', 'Set the end date to scrape', defaultOptions.endDate)
    .option('--run-task-endpoint <runTaskEndpoint>', 'Set the apify run task endpoint', defaultOptions.runTaskEndpoint)
    .parse(process.argv)

async function main() {
    log.setLevel(getLogLevel());

    const inputJson = await jsonfile.readFile(program.inputJsonFile);
    // TODO: make all option names camelCase, so we don't need to transform them
    const options = {
        DRY_RUN: program.dryRun,
        REQUESTS_PER_DAY: program.requestsPerDay,
        REQUEST_DEPTHS_PER_DAY: program.requestDepthsPerDay,
        DAILY: program.daily,
        START_DATE: program.startData,
        END_DATE: program.endDate,
        RUN_TASK_ENDPOINT: program.runTaskEndpoint,
    };

    await apify.scrape(options, { INPUT: inputJson });
}

function getLogLevel() {
    if (program.silent) {
        return log.levels.WARN;
    } else if (program.debug) {
        return log.levels.DEBUG;
    } else {
        return log.levels.INFO;
    }
}

main();


fcli-archive.js
const log = require('loglevel');
const program = require('commander');
const apify = require('apify-etl-lib')();

const defaultOptions = {
    silent: false,
    debug: false,
    downloadDir: 'download',
    rawDataDir: 'raw',
    normalizedDataDir: 'normalized',
    archivedDir: 'archived'
};

program
    .description('Retrieves last run dataset items')
    .option('-s, --silent', 'Set log level to warn', defaultOptions.silent)
    .option('-d, --debug', 'Set log level to debug', defaultOptions.debug)
    .option('--download-dir <downloadDir>', 'Set the download directory to archive', defaultOptions.downloadDir)
    .option('--raw-data-dir <rawDataDir>', 'Set the raw data directory to archive', defaultOptions.rawDataDir)
    .option('--normalized-data-dir <normalizedDataDir>', 'Set the normalized data directory to archive', defaultOptions.normalizedDataDir)
    .option('--archived-dir <archivedDir>', 'Set the destination archived directory', defaultOptions.archivedDir)
    .option('--archived-download-dir <archivedDownloadDir>', 'Set the destination archived download directory')
    .option('--archived-raw-data-dir <archivedRawDataDir>', 'Set the destination archived raw data directory')
    .option('--archived-normalized-data-dir <archivedNormalizedDataDir>', 'Set the destination archived normalized directory')
    .option('--skip-archive-download', 'Skips archiving download directory')
    .parse(process.argv)

async function main() {
    log.setLevel(getLogLevel());

    // TODO: make all option names camelCase, so we don't need to transform them
    const options = {
        ARCHIVED_DIR: program.archivedDir,
        DOWNLOAD_DIR: program.downloadDir,
        RAW_DATA_DIR: program.rawDataDir,
        NORMALIZED_DATA_DIR: program.normalizedDataDir,
        ARCHIVED_DOWNLOAD_DIR: program.archivedDownloadDir,
        ARCHIVED_RAW_DATA_DIR: program.archivedRawDataDir,
        ARCHIVED_NORMALIZED_DATA_DIR: program.archivedNormalizedDataDir,
        SKIP_ARCHIVE_DOWNLOAD: program.skipArchiveDownload,
    };

    await apify.archive(options);
}

function getLogLevel() {
    if (program.silent) {
        return log.levels.WARN;
    } else if (program.debug) {
        return log.levels.DEBUG;
    } else {
        return log.levels.INFO;
    }
}

main();


fcli-retrieve.js
const log = require('loglevel');
const program = require('commander');
const apify = require('apify-etl-lib')();

const defaultOptions = {
    silent: false,
    debug: false,
    rawDataDir: 'raw',
    getDatasetEndpoint: 'https://api.apify.com/v2/actor-tasks/GET_DATASET_ENDPOINT',
    dataFilename: 'data.json'
};

program
    .description('Retrieves last run dataset items')
    .option('-s, --silent', 'Set log level to warn', defaultOptions.silent)
    .option('-d, --debug', 'Set log level to debug', defaultOptions.debug)
    .option('--raw-data-dir <rawDataDir>', 'Set the raw data directory to archive', defaultOptions.rawDataDir)
    .option('--get-dataset-endpoint <getDatasetEndpoint>', 'Set the apify endpoint to retrieve last successful run dataset items', defaultOptions.getDatasetEndpoint)
    .option('--data-file <dataFile>', 'Set the filename to store the data file to', defaultOptions.dataFile)
    .parse(process.argv)

async function main() {
    log.setLevel(getLogLevel());

    // TODO: make all option names camelCase, so we don't need to transform them
    const options = {
        RAW_DATA_DIR: program.rawDataDir,
        GET_DATASET_ENDPOINT: program.rawDataDir,
        DATA_FILE: program.dataFile
    };

    await apify.archive(options);
}

function getLogLevel() {
    if (program.silent) {
        return log.levels.WARN;
    } else if (program.debug) {
        return log.levels.DEBUG;
    } else {
        return log.levels.INFO;
    }
}

main();


fcli-normalize.js
const log = require('loglevel');
const program = require('commander');
const apify = require('apify-etl-lib')();

const defaultOptions = {
    silent: false,
    debug: false,
    maxConcurrency: 10,
    lifeListThumbnailWidth: 120,
    lifeListThumbnailHeight: 77,
    lifeDetailThumbnailWidth: 400,
    lifeDetailThumbnailHeight: 300,
    development: true,
    dataFile: 'example.json',
    downloadDir: 'attachs',
    rawDataDir: 'raw',
    normalizedDataDir: 'normalized',
};

program
    .description('Normalizes data extracted with scraper. Normalizer will parse data json files and populate meaningful data into the transformed data json.')
    .option('-s, --silent', 'Set log level to warn', defaultOptions.silent)
    .option('-d, --debug', 'Set log level to debug', defaultOptions.debug)
    .option('--download-dir <downloadDir>', 'Set the download directory to archive', defaultOptions.downloadDir)
    .option('--raw-data-dir <rawDataDir>', 'Set the raw data directory to archive', defaultOptions.rawDataDir)
    .option('--normalized-data-dir <normalizedDataDir>', 'Set the normalized data directory to archive', defaultOptions.normalizedDataDir)
    .option('-c, --max-concurrency <maxConcurrency>', 'Set max concurrent async requests to process at a time', defaultOptions.maxConcurrency)
    .option('--life-list-thumbnail-width <lifeListThumbnailWidth>', 'Set the width of life list thumbnail', defaultOptions.lifeListThumbnailWidth)
    .option('--list-list-thubmnail-height <lifeListThumbnailHeight>', 'Set the height of life list thumbnail', defaultOptions.lifeListThumbnailHeight)
    .option('--life-detail-thumbnail-width <lifeDetailThumbnailWidth>', 'Set the width of life detail thumbnail', defaultOptions.lifeDetailThumbnailWidth)
    .option('--life-detail-thumbnail-height <lifeDetailThumbnailHeight>', 'Set the height of life detail thumbnail', defaultOptions.lifeDetailThumbnailHeight)
    .option('-D, --development', 'Set development mode to true', defaultOptions.development)
    .option('--data-file <dataFile>', 'Set the data file to process in development', defaultOptions.dataFile)
    .parse(process.argv)

async function main() {
    log.setLevel(getLogLevel());

    const options = {
        maxConcurrency: program.maxConcurrency,
        logLevel: getLogLevel(),
        lifeListThumbnail: {
            width: program.lifeListThumbnailWidth,
            height: program.lifeListThumbnailHeight,
        },
        lifeDetailThumbnail: {
            width: program.lifeDetailThumbnailWidth,
            height: program.lifeDetailThumbnailHeight,
        },
        isDevelopment: program.development,
        dataFile: program.dataFile,
        dir: {
            download: program.downloadDir,
            rawData: program.rawDataDir,
            normalizedData: program.normalizedDataDir,
        }
    };

    await apify.normalize(options);
}

function getLogLevel() {
    if (program.silent) {
        return log.levels.WARN;
    } else if (program.debug) {
        return log.levels.DEBUG;
    } else {
        return log.levels.INFO;
    }
}

main();
